Fallback for storing sensitive stuff when keyring  is not available

kalisp opened the pull request:
Currently on headless Linux nodes (without X11) fails with `keyring.errors.NoKeyringError: No recommended backend was available`. 
This PR adds possibility to use keyring based on encrypted file.

Issue is occurring mostly on render nodes without GUI. Such nodes need to have set environment variables:
- OPENPYPE_MONGO 
- OPENPYPE_SECRET

OPENPYPE_MONGO might be omitted after this PR IF we will figure out how to store this value into new ecrypted file based keyring.

Closes #2060

## Testing notes:
- create new Linux based build (probably for Debian)
- deploy to machine without X
- run `openpype_console`

mkolar commented:
Task linked: [OP-1730 Fallback for storing sensitive stuff when keyring 🔒 is not available](https://app.clickup.com/t/6658547/OP-1730)

iLLiCiTiT commented on igniter/user_settings.py:
```diff
@@ -132,6 +134,28 @@ def delete_item(self, name):
+        from keyrings.cryptfile.cryptfile import CryptFileKeyring
+
+        backends = keyring.get_keyring().backends
+        if backends and \
+                (isinstance(backends[0], EncryptedKeyring) or
+                 isinstance(backends[0], CryptFileKeyring)):
```
```suggestion
        if (
            backends
            and (isinstance(backends[0], EncryptedKeyring)
            or isinstance(backends[0], CryptFileKeyring)
        ):
```

iLLiCiTiT commented:
So instead of setting env variable `OPENPYPE_MONGO` it is required to set `OPENPYPE_SECRET`? How do I actually set the mongo url?

> OPENPYPE_MONGO might be omitted after this PR IF we will figure out how to store this value into new ecrypted file based keyring.

Maybe for farm publishing but not for other processes.
- on artists machines e.g. Maya can't use keyring at all so would not have access to mongo URL.
- ftrack event server is actually 1 main process and 3 child processes - the children processes must use same mongo url as main process no matter what is set in keyring

kalisp commented:
I added comment about OPENPYPE_MONGO situation is just for render machines. I highlighted it there as an option if someone would be really squeamish about setting OPENPYPE_MONGO as an env var. In that case we would need to create some script which would be run on each machine to insert that value info keyring after deployment of OP (which is not really practical, but maybe someone will require that).

I guess this is mainly as a fallback option to limit pesky issues on Deadline nodes and provide same "interface" for secrets on all environments.

antirotor commented:
This is also useful in various headless uses - event server, some batch processing in the future, and so on. I think we could add command to openpype to store something in that keyring like:

```sh
openpype_console store-value OPENPYPE_MONGO=foo
```
(along with `get-value`?)

Or maybe even something more common, for getting/setting various kinds of values form OpenPype?

```sh
openpype_console set-value settings://system/something=something
openpype_console set-value keyring://file/environment/OPENPYPE_MONGO=foo
openpype_console set-value keyring://system/openpype/key=value
```

kalisp commented:
Implemented commands:
-  `openpype_console setvalue keyring://file/environment/OPENPYPE_MONGO=foo`
- `openpype_console getvalue keyring://file/environment/OPENPYPE_MONGO`

antirotor commented:
well there are few issues...

`setvalue keyring://foo/bar=111` and then `getvalue keyring://foo/bar` will return nothing for me (it is without `file` part of URL.

![image](https://user-images.githubusercontent.com/33513211/148764415-22b04200-b77a-44ac-9ce4-1cc0cb7c5461.png)

it should return something about invalid keyring resource?

getting non-existent value will produce traceback - I think we should handle it more gracefully (non-zero exit code and simple message about accessing invalid key.

same thing if you trash URI schema and write `foo://` instead of `keyring://` - this will produce traceback.

kalisp commented:
Fixed, added unit tests.

This still doesn't solve an issue of setting of `OPENPYPE_MONGO` as before it gets to `setvalue` method it already checks if Mongo is configured.

kalisp commented:
Explicit check for `setvalue` command for `OPENPYPE_MONGO` was added into `start.py`

iLLiCiTiT commented on openpype/pype_commands.py:
```diff
@@ -403,7 +403,10 @@ def get_value(get_url, project=None):
         if general_location == "keyring":
             registry = OpenPypeSecureRegistry(path_part)
-            return registry.get_item(key)
+            try:
```
Should we skip the exception and pretend that nothing happened? Result will be None which would look like the value is not set. But we don't know if is set because it can't be accessed.

kalisp replied:
`get_secure_value` in lib throws an exception if key not found.
Cli catches the exception and prints it on stdout.

iLLiCiTiT commented on openpype/pype_commands.py:
```diff
@@ -384,6 +384,30 @@ def run_tests(self, folder, mark, pyargs,
         pytest.main(args)
 
+    @staticmethod
+    def set_value(set_url, project=None):
```
I would change the function names to `get_secure_value` and `set_secure_value` as `get_value` won't tell much. Also these functions should be just wrappers for functions in lib that contain the logic so we can call it from OpenPype directly if needed like when MongoURL is changed from local settings.

kalisp replied:
Sounds reasonable, changed.

iLLiCiTiT commented on start.py:
```diff
@@ -568,6 +568,11 @@ def _determine_mongodb() -> str:
 
     """
+    # special case when OPENPYPE_MONGO is being set into keyring
+    if "setvalue" in sys.argv:
```
```suggestion
    if "setsecurevalue" in sys.argv:
```

iLLiCiTiT commented on start.py:
```diff
@@ -568,6 +568,11 @@ def _determine_mongodb() -> str:
+    # special case when OPENPYPE_MONGO is being set into keyring
+    if "setvalue" in sys.argv:
+        key, value = sys.argv[-1].split("=")
+        if "OPENPYPE_MONGO" in key:
```
```suggestion
        if "OPENPYPE_MONGO" == key:
```

iLLiCiTiT commented on website/docs/admin_use.md:
````diff
@@ -89,6 +89,18 @@ save it in secure way to your systems keyring - on Windows it is **Credential Ma
+
+Then as a first set `OPENPYPE_MONGO` into a keyring:
+```shell
+openpype_console setvalue keyring://file/environment/OPENPYPE_MONGO=YOUR_MONGO_URL
````
```suggestion
openpype_console setsecurevalue keyring://file/environment/OPENPYPE_MONGO=YOUR_MONGO_URL
```

iLLiCiTiT commented on openpype/lib/local_settings.py:
```diff
@@ -35,6 +35,7 @@ def wrapper(*args, **kwargs):
 )
 
 from .import validate_mongo_connection
+from igniter.tools import (parse_get_url, parse_set_url)
```
You can't import from igniter in `openpype.lib`. Igniter is available only in openpype process but not in any python host (Maya, Nuke, Houdini, etc.).

iLLiCiTiT replied:
Right now you have to move them to openpype and have copy of `parse_set_url` in igniter. We should create some folder in igniter that is added to pythonpath and sys.path where functions used for both build and code are.

NOTE: This PR will require new build.

kalisp replied:
This is blocking issue for this PR to be merged. Currently is not possible to share code between igniter and OP.

mkolar commented:
@kalisp @antirotor @iLLiCiTiT guys what did we get stuck on here? any chance of resolving it?

mkolar commented:
@kalisp I'm going to close this considering we'll need to sort it in a more robust way with the new client structure in v4.
